File: MacApp/Sources/VTubeLinkService/DataMapper.swift
import Foundation
import VTubeLinkShared

struct VTStudioParam: Encodable {
    let id: String
    let value: Float
}

class DataMapper {
    static func buildParamsDict(from data: CapturedData) -> [VTStudioParam] {
        var params: [VTStudioParam] = []
        
        let blendshapes = data.blendshapes
        
        func bs(_ key: String) -> Float {
            return blendshapes[key] ?? 0.0
        }
        
        // VTubeStudio Default
        params.append(VTStudioParam(id: "FacePositionX", value: data.headPositionX * Constants.facePositionXRatio))
        params.append(VTStudioParam(id: "FacePositionY", value: data.headPositionY * Constants.facePositionYRatio))
        params.append(VTStudioParam(id: "FacePositionZ", value: -data.headPositionZ * Constants.facePositionZRatio))
        
        params.append(VTStudioParam(id: "FaceAngleX", value: data.headRotationY * Constants.faceAngleXRatio))
        params.append(VTStudioParam(id: "FaceAngleY", value: -data.headRotationX * Constants.faceAngleYRatio))
        params.append(VTStudioParam(id: "FaceAngleZ", value: -data.headRotationZ * Constants.faceAngleZRatio))
        
        let mouthSmileLeft = bs("mouthSmile_L")
        let mouthSmileRight = bs("mouthSmile_R")
        let mouthShrugLower = bs("mouthShrugLower")
        let browDownLeft = bs("browDown_L")
        let browDownRight = bs("browDown_R")
        let jawOpen = bs("jawOpen")
        let mouthClose = bs("mouthClose")
        
        let smileInner1 = max(mouthSmileLeft + mouthSmileRight - 0.2, 0.0)
        let smileInner2 = pow(max(mouthShrugLower - 0.4, 0.0), 1.0) * 1.0
        let browDownAvg = (browDownLeft + browDownRight) / 2.0
        let mouthOpenFactor = (jawOpen - mouthClose) * 0.15
        let smileInner3 = pow(max(browDownAvg - 0.3 + (0.08 + mouthOpenFactor), 0.0), 0.4) * 1.5
        
        let mouthSmileVal = ((smileInner1 - smileInner2 - smileInner3) * Constants.mouthSmileRatio) / 2.0 + 0.5
        params.append(VTStudioParam(id: "MouthSmile", value: mouthSmileVal))
        
        params.append(VTStudioParam(id: "MouthOpen", value: (jawOpen - mouthClose) * Constants.mouseOpenRatio))
        params.append(VTStudioParam(id: "Brows", value: bs("browInnerUp") * Constants.browsRatio))
        
        let tongueOut = bs("tongueOut")
        params.append(VTStudioParam(id: "TongueOut", value: tongueOut < Constants.tongueOutRatio ? 0.0 : 1.0))
        
        let eyeOpenLeft = (1.0 - bs("eyeBlink_L")) * Constants.eyeOpenRatio - (Constants.eyeOpenRatio - 1.0)
        params.append(VTStudioParam(id: "EyeOpenLeft", value: eyeOpenLeft))
        
        let eyeOpenRight = (1.0 - bs("eyeBlink_R")) * Constants.eyeOpenRatio - (Constants.eyeOpenRatio - 1.0)
        params.append(VTStudioParam(id: "EyeOpenRight", value: eyeOpenRight))
        
        params.append(VTStudioParam(id: "EyeLeftX", value: (bs("eyeLookIn_L") - bs("eyeLookOut_L")) * Constants.eyeRotationRatio))
        params.append(VTStudioParam(id: "EyeLeftY", value: (bs("eyeLookUp_L") - bs("eyeLookDown_L")) * Constants.eyeRotationRatio))
        params.append(VTStudioParam(id: "EyeRightX", value: (bs("eyeLookOut_R") - bs("eyeLookIn_R")) * Constants.eyeRotationRatio))
        params.append(VTStudioParam(id: "EyeRightY", value: (bs("eyeLookUp_R") - bs("eyeLookDown_R")) * Constants.eyeRotationRatio))
        
        params.append(VTStudioParam(id: "CheekPuff", value: bs("cheekPuff") * Constants.cheekPuffRatio))
        
        let mouthRollLower = bs("mouthRollLower")
        params.append(VTStudioParam(id: "FaceAngry", value: (mouthRollLower * mouthShrugLower) < Constants.faceAngryRatio ? 0.0 : 1.0))
        
        let browOuterUpLeft = bs("browOuterUp_L")
        params.append(VTStudioParam(id: "BrowLeftY", value: ((browOuterUpLeft - browDownLeft) * Constants.browRightYRatio + 1.0) / 2.0))
        
        let browOuterUpRight = bs("browOuterUp_R")
        params.append(VTStudioParam(id: "BrowRightY", value: ((browOuterUpRight - browDownRight) * Constants.browRightYRatio + 1.0) / 2.0))
        
        params.append(VTStudioParam(id: "MouthX", value: (bs("mouthLeft") - bs("mouthRight")) * Constants.mouthXRatio))
        
        // Custom Params (ARKit) mapped from blendshapes 1-1 to customParams names
        let nameMappings: [String: String] = [
            "EyeBlinkLeft": "eyeBlink_L", "EyeLookDownLeft": "eyeLookDown_L", "EyeLookInLeft": "eyeLookIn_L",
            "EyeLookOutLeft": "eyeLookOut_L", "EyeLookUpLeft": "eyeLookUp_L", "EyeSquintLeft": "eyeSquint_L",
            "EyeWideLeft": "eyeWide_L", "EyeBlinkRight": "eyeBlink_R", "EyeLookDownRight": "eyeLookDown_R",
            "EyeLookInRight": "eyeLookIn_R", "EyeLookOutRight": "eyeLookOut_R", "EyeLookUpRight": "eyeLookUp_R",
            "EyeSquintRight": "eyeSquint_R", "EyeWideRight": "eyeWide_R", "JawForward": "jawForward",
            "JawLeft": "jawLeft", "JawRight": "jawRight", "JawOpen": "jawOpen", "MouthClose": "mouthClose",
            "MouthFunnel": "mouthFunnel", "MouthPucker": "mouthPucker", "MouthLeft": "mouthLeft", "MouthRight": "mouthRight",
            "MouthSmileLeft": "mouthSmile_L", "MouthSmileRight": "mouthSmile_R", "MouthFrownLeft": "mouthFrown_L",
            "MouthFrownRight": "mouthFrown_R", "MouthDimpleLeft": "mouthDimple_L", "MouthDimpleRight": "mouthDimple_R",
            "MouthStretchLeft": "mouthStretch_L", "MouthStretchRight": "mouthStretch_R", "MouthRollLower": "mouthRollLower",
            "MouthRollUpper": "mouthRollUpper", "MouthShrugLower": "mouthShrugLower", "MouthShrugUpper": "mouthShrugUpper",
            "MouthPressLeft": "mouthPress_L", "MouthPressRight": "mouthPress_R", "MouthLowerDownLeft": "mouthLowerDown_L",
            "MouthLowerDownRight": "mouthLowerDown_R", "MouthUpperUpLeft": "mouthUpperUp_L", "MouthUpperUpRight": "mouthUpperUp_R",
            "BrowDownLeft": "browDown_L", "BrowDownRight": "browDown_R", "BrowInnerUp": "browInnerUp",
            "BrowOuterUpLeft": "browOuterUp_L", "BrowOuterUpRight": "browOuterUp_R", "CheekPuff": "cheekPuff",
            "CheekSquintLeft": "cheekSquint_L", "CheekSquintRight": "cheekSquint_R", "NoseSneerLeft": "noseSneer_L",
            "NoseSneerRight": "noseSneer_R", "TongueOut": "tongueOut"
        ]
        
        for customParam in Constants.customParams {
            if let mappedKey = nameMappings[customParam] {
                params.append(VTStudioParam(id: customParam, value: bs(mappedKey)))
            } else {
                params.append(VTStudioParam(id: customParam, value: 0.0))
            }
        }
        
        return params
    }
}

File: WinApp/DataMapper.cs
using System;
using System.Collections.Generic;

namespace VTubeLink
{
    public static class DataMapper
    {
        public static List<VTStudioParam> BuildParamsDict(CapturedData data)
        {
            var parameters = new List<VTStudioParam>();
            var blendshapes = data.Blendshapes;

            float Bs(string key) => blendshapes.TryGetValue(key, out var val) ? val : 0f;

            // VTubeStudio Default
            parameters.Add(new VTStudioParam { Id = "FacePositionX", Value = data.HeadPositionX * Constants.FacePositionXRatio });
            parameters.Add(new VTStudioParam { Id = "FacePositionY", Value = data.HeadPositionY * Constants.FacePositionYRatio });
            parameters.Add(new VTStudioParam { Id = "FacePositionZ", Value = -data.HeadPositionZ * Constants.FacePositionZRatio });

            parameters.Add(new VTStudioParam { Id = "FaceAngleX", Value = data.HeadRotationY * Constants.FaceAngleXRatio });
            parameters.Add(new VTStudioParam { Id = "FaceAngleY", Value = -data.HeadRotationX * Constants.FaceAngleYRatio });
            parameters.Add(new VTStudioParam { Id = "FaceAngleZ", Value = -data.HeadRotationZ * Constants.FaceAngleZRatio });

            float mouthSmileLeft = Bs("mouthSmile_L");
            float mouthSmileRight = Bs("mouthSmile_R");
            float mouthShrugLower = Bs("mouthShrugLower");
            float browDownLeft = Bs("browDown_L");
            float browDownRight = Bs("browDown_R");
            float jawOpen = Bs("jawOpen");
            float mouthClose = Bs("mouthClose");

            float smileInner1 = Math.Max(mouthSmileLeft + mouthSmileRight - 0.2f, 0.0f);
            float smileInner2 = (float)Math.Pow(Math.Max(mouthShrugLower - 0.4f, 0.0f), 1.0f) * 1.0f;
            float browDownAvg = (browDownLeft + browDownRight) / 2.0f;
            float mouthOpenFactor = (jawOpen - mouthClose) * 0.15f;
            float smileInner3 = (float)Math.Pow(Math.Max(browDownAvg - 0.3f + (0.08f + mouthOpenFactor), 0.0f), 0.4f) * 1.5f;

            float mouthSmileVal = ((smileInner1 - smileInner2 - smileInner3) * Constants.MouthSmileRatio) / 2.0f + 0.5f;
            parameters.Add(new VTStudioParam { Id = "MouthSmile", Value = mouthSmileVal });

            parameters.Add(new VTStudioParam { Id = "MouthOpen", Value = (jawOpen - mouthClose) * Constants.MouseOpenRatio });
            parameters.Add(new VTStudioParam { Id = "Brows", Value = Bs("browInnerUp") * Constants.BrowsRatio });

            float tongueOut = Bs("tongueOut");
            parameters.Add(new VTStudioParam { Id = "TongueOut", Value = tongueOut < Constants.TongueOutRatio ? 0.0f : 1.0f });

            float eyeOpenLeft = (1.0f - Bs("eyeBlink_L")) * Constants.EyeOpenRatio - (Constants.EyeOpenRatio - 1.0f);
            parameters.Add(new VTStudioParam { Id = "EyeOpenLeft", Value = eyeOpenLeft });

            float eyeOpenRight = (1.0f - Bs("eyeBlink_R")) * Constants.EyeOpenRatio - (Constants.EyeOpenRatio - 1.0f);
            parameters.Add(new VTStudioParam { Id = "EyeOpenRight", Value = eyeOpenRight });

            parameters.Add(new VTStudioParam { Id = "EyeLeftX", Value = (Bs("eyeLookIn_L") - Bs("eyeLookOut_L")) * Constants.EyeRotationRatio });
            parameters.Add(new VTStudioParam { Id = "EyeLeftY", Value = (Bs("eyeLookUp_L") - Bs("eyeLookDown_L")) * Constants.EyeRotationRatio });
            parameters.Add(new VTStudioParam { Id = "EyeRightX", Value = (Bs("eyeLookOut_R") - Bs("eyeLookIn_R")) * Constants.EyeRotationRatio });
            parameters.Add(new VTStudioParam { Id = "EyeRightY", Value = (Bs("eyeLookUp_R") - Bs("eyeLookDown_R")) * Constants.EyeRotationRatio });

            parameters.Add(new VTStudioParam { Id = "CheekPuff", Value = Bs("cheekPuff") * Constants.CheekPuffRatio });

            float mouthRollLower = Bs("mouthRollLower");
            parameters.Add(new VTStudioParam { Id = "FaceAngry", Value = (mouthRollLower * mouthShrugLower) < Constants.FaceAngryRatio ? 0.0f : 1.0f });

            float browOuterUpLeft = Bs("browOuterUp_L");
            parameters.Add(new VTStudioParam { Id = "BrowLeftY", Value = ((browOuterUpLeft - browDownLeft) * Constants.BrowRightYRatio + 1.0f) / 2.0f });

            float browOuterUpRight = Bs("browOuterUp_R");
            parameters.Add(new VTStudioParam { Id = "BrowRightY", Value = ((browOuterUpRight - browDownRight) * Constants.BrowRightYRatio + 1.0f) / 2.0f });

            parameters.Add(new VTStudioParam { Id = "MouthX", Value = (Bs("mouthLeft") - Bs("mouthRight")) * Constants.MouthXRatio });

            // Custom Params (ARKit) mapped from blendshapes 1-1 to customParams names
            var nameMappings = new Dictionary<string, string>
            {
                { "EyeBlinkLeft", "eyeBlink_L" }, { "EyeLookDownLeft", "eyeLookDown_L" }, { "EyeLookInLeft", "eyeLookIn_L" },
                { "EyeLookOutLeft", "eyeLookOut_L" }, { "EyeLookUpLeft", "eyeLookUp_L" }, { "EyeSquintLeft", "eyeSquint_L" },
                { "EyeWideLeft", "eyeWide_L" }, { "EyeBlinkRight", "eyeBlink_R" }, { "EyeLookDownRight", "eyeLookDown_R" },
                { "EyeLookInRight", "eyeLookIn_R" }, { "EyeLookOutRight", "eyeLookOut_R" }, { "EyeLookUpRight", "eyeLookUp_R" },
                { "EyeSquintRight", "eyeSquint_R" }, { "EyeWideRight", "eyeWide_R" }, { "JawForward", "jawForward" },
                { "JawLeft", "jawLeft" }, { "JawRight", "jawRight" }, { "JawOpen", "jawOpen" }, { "MouthClose", "mouthClose" },
                { "MouthFunnel", "mouthFunnel" }, { "MouthPucker", "mouthPucker" }, { "MouthLeft", "mouthLeft" }, { "MouthRight", "mouthRight" },
                { "MouthSmileLeft", "mouthSmile_L" }, { "MouthSmileRight", "mouthSmile_R" }, { "MouthFrownLeft", "mouthFrown_L" },
                { "MouthFrownRight", "mouthFrown_R" }, { "MouthDimpleLeft", "mouthDimple_L" }, { "MouthDimpleRight", "mouthDimple_R" },
                { "MouthStretchLeft", "mouthStretch_L" }, { "MouthStretchRight", "mouthStretch_R" }, { "MouthRollLower", "mouthRollLower" },
                { "MouthRollUpper", "mouthRollUpper" }, { "MouthShrugLower", "mouthShrugLower" }, { "MouthShrugUpper", "mouthShrugUpper" },
                { "MouthPressLeft", "mouthPress_L" }, { "MouthPressRight", "mouthPress_R" }, { "MouthLowerDownLeft", "mouthLowerDown_L" },
                { "MouthLowerDownRight", "mouthLowerDown_R" }, { "MouthUpperUpLeft", "mouthUpperUp_L" }, { "MouthUpperUpRight", "mouthUpperUp_R" },
                { "BrowDownLeft", "browDown_L" }, { "BrowDownRight", "browDown_R" }, { "BrowInnerUp", "browInnerUp" },
                { "BrowOuterUpLeft", "browOuterUp_L" }, { "BrowOuterUpRight", "browOuterUp_R" }, { "CheekPuff", "cheekPuff" },
                { "CheekSquintLeft", "cheekSquint_L" }, { "CheekSquintRight", "cheekSquint_R" }, { "NoseSneerLeft", "noseSneer_L" },
                { "NoseSneerRight", "noseSneer_R" }, { "TongueOut", "tongueOut" }
            };

            foreach (var customParam in Constants.CustomParams)
            {
                if (nameMappings.TryGetValue(customParam, out var mappedKey))
                {
                    parameters.Add(new VTStudioParam { Id = customParam, Value = Bs(mappedKey) });
                }
                else
                {
                    parameters.Add(new VTStudioParam { Id = customParam, Value = 0.0f });
                }
            }

            return parameters;
        }
    }
}
